CARDS-1347: Evaluate computed fields serverside

acrowthe opened the pull request:
- If a computed question is not filled out, compute and fill out the answer
  - If the answers required to compute a question is not available,
    does not create the question
- Creates section tree as required to insert questions
  - Assumes all recurrent sections should exist `initialNumberOfInstances` times
  - Assumes all conditional sections should have answers
  
  Testing:
  The form `Serverside Computed Field Test` has been created in the test runmode containing computed fields in a conditional section. If the question `Display computed section` is unanswered or `No`, then the web browser will not compute these questions but the backend will.

acrowthe committed:
CARDS-1347: Evaluate computed fields serverside
- If a computed question is not filled out, compute and fill out the answer
  - If the answers required to compute a question is not available,
    does not create the question
- Creates section tree as required to insert questions
  - Assumes all recurrent sections should exist `initialNumberOfInstances` times
  - Assumes all conditional sections should have answers

sdumitriu committed:
CARDS-1347: Evaluate computed fields serverside
Better way to evaluate an expression with input parameters: use bindings

sdumitriu commented on modules/data-entry/src/main/java/io/uhndata/cards/ComputedEditor.java:
```diff
@@ -0,0 +1,512 @@
+ * @version $Id$
+ */
+@SuppressWarnings("checkstyle:ClassFanOutComplexity")
+public class ComputedEditor extends DefaultEditor
```
Better name: `ComputedAnswersEditor`.

sdumitriu commented on modules/data-entry/src/main/java/io/uhndata/cards/ComputedEditor.java:
```diff
@@ -0,0 +1,512 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.uhndata.cards;
```
Should be in `io.uhndata.cards.internal`.

sdumitriu commented on modules/data-entry/src/main/java/io/uhndata/cards/ComputedEditor.java:
```diff
@@ -0,0 +1,512 @@
+            engine.eval(expression);
+            Invocable invocable = (Invocable) engine;
+            Object result = invocable.invokeFunction("compute",
+                parsedExpression.getInputValues().toArray(new Object[0]));
```
This is hacky. A better way to evaluate an expression with provided parameters is to use `ScriptEngine.eval(String, Bindings)`.

sdumitriu reviewed (commented):
Two things that are problematic:

- In the browser, although numbers are all double, they are printed in a prettier way, so `1 + 2 + 3 => 6`, but in Java, the result is `6.0`. That makes, for example, the GAD7 score in proms mode be the verbose `### 18.0 points`. I think manually pretifying double results should be an easy fix for this.
- In React it's easy to have the computed the `gad7_result` depend on the computed `gad7_score`, since every answer depends on the form context, and computing `gad7_score` updates the context, which causes `gad7_result` to be re-evaluated. In the current backend implementation, by the way the questions are ordered, `gad7_result` is evaluated first, a missing dependency is encountered, so no answer is created for it. After `gad7_score` is computed, no attempt is made to recompute `gad7_result`. In other words, the first time a form is evaluated, no answer for `gad7_result` is created, but the second time the form is changed, it will be computed, but with the previous value of `gad7_score`. This might be worked around by reordering the questions in the form (it was done like this to avoid a styling issue), but I wonder if we can do a dependency ordering in the backend to always evaluate questions in the dependency order, but still create nodes in the expected order.

sdumitriu committed:
CARDS-1347: Evaluate computed fields serverside
Moved editor classes to the `internal` subpackage

marta- commented:
> This might be worked around by reordering the questions in the form (it was done like this to avoid a styling issue), but I wonder if we can do a dependency ordering in the backend to always evaluate questions in the dependency order, but still create nodes in the expected order.

I'll get right on fixing the styling issue in another branch. Regardless, dependency order sounds like a smart move.

sdumitriu commented:
> If the question Display computed section is unanswered or No, then the web browser will not compute these questions but the backend will.

I think that is not the correct behavior. The section should not be there if the conditions for it are not met.

sdumitriu commented:
Actually the questionnaire order doesn't matter, since the questions are in a pseudorandom `HashMap`, so we must do a dependency ordering anyway.

sdumitriu committed:
CARDS-1347: Evaluate computed fields serverside
Improve result formatting

sdumitriu committed:
CARDS-1347: Evaluate computed fields serverside
Turn inner classes into static final classes, since they don't need to be instance classes or be modifiable

sdumitriu committed:
CARDS-1347: Evaluate computed fields serverside
Bugfix: default values should be used also when the answer exists but has no value, not just when the answer doesn't exist

sdumitriu committed:
CARDS-1347: Evaluate computed fields serverside
Cleanup: no need to keep track of the versionable ancestor

sdumitriu committed:
CARDS-1347: Evaluate computed fields serverside
Compute all questions that have not been modified by the current changeset, even if already present

sdumitriu committed:
CARDS-1347: Evaluate computed fields serverside
Refactoring: move expression evaluation code into a separate subclass

sdumitriu committed:
CARDS-1347: Evaluate computed fields serverside
Refactoring: first create all missing nodes, then evaluate answers.
This is needed to be able to compute answers in order of their interdependencies.

sdumitriu committed:
[misc] Improved the description of the AnswerSection node type in the CND

sdumitriu committed:
CARDS-1347: Evaluate computed fields serverside
- Moved code from io.uhndata.cards.internal to io.uhndata.cards.dataentry.internal
- Extracted new public API for working with forms (io.uhndata.cards.dataentry.api.FormUtils)
- Extracted new public API for working with questionnaires (io.uhndata.cards.dataentry.api.QuestionnaireUtils)
- Extracted the code for evaluating expressions into a separate class file, should be turned into an actual API next
- Sorted computed quesions by their inter-dependencies to avoid missing or deprecated computed answers

sdumitriu committed:
CARDS-1347: Evaluate computed fields serverside
More code documentation

sdumitriu committed:
CARDS-1347: Evaluate computed fields serverside
Avoid infinite recursion if there's a circular computed question interdependency

sashaandjic reviewed (approved):
tested

marta- commented on modules/data-entry/src/main/java/io/uhndata/cards/dataentry/api/QuestionnaireUtils.java:
```diff
@@ -0,0 +1,135 @@
+    String QUESTION_NODETYPE = "cards:Question";
+
+    /** The Sling resource type for a Question. */
+    String QUESTION_RESOURCE = "cards/Questin";
```
?

sdumitriu committed:
[misc] Fixed typo

sashaandjic reviewed (approved):
re-tested

acrowthe commented:
Refactor looks good
